README.md
<p align="center">
<img width="420" src="assets/logo.png"/>
</p>

[![Build Status](https://travis-ci.org/andreamazz/AMPopTip.svg)](https://travis-ci.org/andreamazz/AMPopTip)
[![codecov](https://codecov.io/gh/andreamazz/AMPopTip/branch/master/graph/badge.svg)](https://codecov.io/gh/andreamazz/AMPopTip)
[![CocoaPods](https://cocoapod-badges.herokuapp.com/v/AMPopTip/badge.svg)](http://cocoapods.org/?q=ampoptip)
[![Docs](https://img.shields.io/cocoapods/metrics/doc-percent/AMPopTip.svg)](http://cocoadocs.org/docsets/AMPopTip)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
![Swift 4.2](https://img.shields.io/badge/swift-4.2-orange.svg)
[![Join the chat at https://gitter.im/andreamazz/AMPopTip](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/andreamazz/AMPopTip?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Animated popover that pops out of a frame. You can specify the direction of the popover and the arrow that points to its origin. Color, border radius and font can be easily customized.
This popover can be used to leave subtle hints about your UI and provide fun looking onboarding popups.  

<p align="center">
<a href='https://appetize.io/app/g7x2pwb9jtxdpz9yjb2p5wzk6g' alt='Live demo'>
<img width="150" height="75" src="assets/demo-button.png"/>
</a>
</p>

# Screenshot
![AMPopTip](https://raw.githubusercontent.com/andreamazz/AMPopTip/master/assets/screenshot.gif)

### Versioning notes

With version `2.0.0` the library was re-written in Swift, and the API was slightly updated. Checkout version `1.5.x` for the previous Objective-C implementation.

Version `3.0.0` introduces Swift 4 support, `3.5.0` Swift 4.2.  

# Setup with CocoaPods
* Add ```pod 'AMPopTip'``` to your ```Podfile```
* Run ```pod install```
* Run ```open App.xcworkspace```

# Setup with Carthage
* Add ```github "andreamazz/AMPopTip"```
* Run ```carthage update```
* Add ```AMPopTip.framework``` in your project

# Usage
The API is fairly straight forward, you can show and hide the popover at any time.

## Showing the popover
You must specify the text that you want to display alongside the popover direction, its max width, the view that will contain it and the frame of the view that the popover's arrow will point to.

```swift
let popTip = PopTip()
popTip.show(text: "Hey! Listen!", direction: .up, maxWidth: 200, in: view, from: someView.frame)
```

You can also display the popover in the center, with no arrow, in this case the `from` can be the whole view:
```swift
popTip.show(text: "Hey! Listen!", direction: .none, maxWidth: 200, in: view, from: view.frame)
```

## Coordinate system
Please note that the frame you are intended to provide needs to refer to the absolute coordinate system of the view you are presenting the popover in. This means that if you are presenting the popover in a view, pointing to a nested subview, you'll need to convert its frame using UIKit's `convertRect(_:toView:)`. Read the reference [here](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIView_Class/#//apple_ref/occ/instm/UIView/convertRect:toView:).

## Showing a custom view
You can provide a custom view that will be wrapped in the PopTip and presented.

```swift
let customView = UIView(frame: CGRect(x: 0, y: 0, width: 100, height: 100))
// Configure your view
popTip.show(customView: customView, direction: .down, in: view, from: someView.frame)
```

## Dismissing the popover

You can hide the popover by calling:
```swift
popTip.hide()
```

Or you can specify the duration of the popover:
```swift
popTip.show(text: "Hey! Listen!", direction: .up, maxWidth: 200, in: view, from: someView.frame, duration: 3)
```

You can also let the user dismiss the popover by tapping on it:
```swift
popTip.shouldDismissOnTap = true
```

You can add a block that will be fired when the user taps the PopTip...
```swift
popTip.tapHandler = { popTip in
print("\(popTip) tapped")
}
```

... when the popover is shown...
```swift
popTip.appearHandler = { popTip in
print("\(popTip) appeared")
};
```

... or when the popover is dismissed:
```swift
popTip.dismissHandler = { popTip in
print("\(popTip) dismissed")
}

popTip.tapOutsideHandler = { _ in
print("tap outside")
}

popTip.swipeOutsideHandler = { _ in
print("swipe outside")
}
```

# Updating the PopTip

You can update the text, attributed text, or custom view to a PopTip already visible:

```swift
popTip.update(text: "New string")
popTip.update(attributedText: someAttributedString)
popTip.update(customView: someView)
```

The position can also be changed by updating the `from` property:

```swift
let here = CGRect(x: 100, y: 100, width: 10, height: 10)
let there = CGRect(x: 400, y: 400, width: 10, height: 10)

popTip.show(text: "Hey! Listen!", direction: .up, maxWidth: 200, in: view, from: here)
DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
popTip.from = there
}
```

# Custom entrance animation

You can choose which animation should be performed when the popTip is displayed:
```swift
popTip.entranceAnimation = .scale;
```

Available animations:
```swift
PopTipEntranceAnimation.scale,
PopTipEntranceAnimation.transition,
PopTipEntranceAnimation.none,
PopTipEntranceAnimation.custom
```

## PopTipEntranceAnimation.custom

You can provide your own animation block when using `PopTipEntranceAnimation.custom`:
```swift
popTip.entranceAnimationHandler = { [weak self] completion in
guard let `self` = self else { return }
self.popTip.transform = CGAffineTransform(rotationAngle: 0.3)
UIView.animate(withDuration: 0.5, animations: {
self.popTip.transform = .identity
}, completion: { (_) in
completion()
})
}
```
This sample makes the PopTip rotate on entrance. Make sure to call the completion block when the animation is done. Also note that the animation is fired as soon as the PopTip is added as subview.

# Action animations
Action animations are subtle animations that can be performed to get the user's attention.
Set your preferred animation:
```swift
popTip.actionAnimation = .bounce()
```

Available animations:
```swift
PopTipActionAnimation.bounce,
PopTipActionAnimation.float,
PopTipActionAnimation.pulse,
PopTipActionAnimation.none
```
The animation is fired as soon as the popover enters the scene and completes its entrance animation, if `startActionAnimationOnShow` is set to true.

## Customize the animations

You can pass a custom value as an associated value to customize the action animation:

```swift
popTip.actionAnimation = .bounce(16) // This will bounce for 16px instead of the default value
```

![AMPopTip bounce](assets/bounce_effect.gif)

# Customizing the arrow position

The arrow is centered by default, and moves to avoid the edge of the screen. You can manually change the offset from the center using the `bubbleOffset` property.

# Customization

Use the appearance proxy to customize the popover before creating the instance, or just use its public properties:
```swift
textColor = <#UIColor#>;
textAlignment = <#NSTextAlignment#>
bubbleColor = <#UIColor#>
borderColor = <#UIColor#>
borderWidth = <#CGFloat#>
cornerRadius = <#CGFloat#> // Popover's border radius
rounded = <#Bool#> // If set to YES the radius will equal frame.height / 2
offset = <#CGFloat#> // Offset between the popover and the origin
font = <#UIFont#>
padding = <#CGFloat#>
edgeInsets = <#UIEdgeInsets#>
arrowSize = <#CGSize#>
animationIn = <#TimeInterval#>
animationOut = <#TimeInterval#>
delayIn = <#TimeInterval#>
delayOut = <#TimeInterval#>
entranceAnimation = <#PopTipEntranceAnimation#>
actionAnimation = <#PopTipActionAnimation#>
actionAnimationIn = <#TimeInterval#>
actionAnimationOut = <#TimeInterval#>
actionDelayIn = <#TimeInterval#>
actionDelayOut = <#TimeInterval#>
edgeMargin = <#CGFloat#>
bubbleOffset = <#CGFloat#> // Offset between the bubble and the arrow
arrowRadius = <#CGFloat#>
shadowOpacity = <#Float#>
shadowRadius = <#Float#>
shadowOffset = <#CGSize#>
shadowColor = <#UIColor#>
```

# Author
[Andrea Mazzini](https://twitter.com/theandreamazz). I'm available for freelance work, feel free to contact me.

Want to support the development of [these free libraries](https://cocoapods.org/owners/734)? Buy me a coffee ☕️ via [Paypal](https://www.paypal.me/andreamazzini).  

# Contributors
Thanks to [everyone](https://github.com/andreamazz/AMPopTip/graphs/contributors) kind enough to submit a pull request.

# MIT License

Copyright (c) 2017 Andrea Mazzini. All rights reserved.

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`AMPopTip` adheres to [Semantic Versioning](http://semver.org/).

- `3.5.2` Releases - [3.5.2](#352)
- `3.5.1` Releases - [3.5.1](#351)
- `3.5.0` Releases - [3.5.0](#350)  
- `3.4.0` Releases - [3.4.0](#340)  
- `3.3.0` Releases - [3.3.0](#330)  
- `3.2.0` Releases - [3.2.0](#320) | [3.2.1](#321)  
- `3.1.0` Releases - [3.1.0](#310) | [3.1.1](#311) | [3.1.2](#312) | [3.1.3](#313) | [3.1.4](#314)  
- `3.0.0` Releases - [3.0.0](#300) | [3.0.1](#301) | [3.0.2](#302) | [3.0.3](#303)  
- `2.1.0` Releases - [2.1.0](#210) | [2.1.1](#211) | [2.1.2](#212) | [2.1.3](#213) | [2.1.4](#214) | [2.1.5](#215) | [2.1.6](#216) | [2.1.7](#217)  
- `2.0.0` Releases - [2.0.0](#200) | [2.0.1](#201)  
- `1.5.0` Releases - [1.5.0](#150) | [1.5.1](#151)  [1.5.2](#152)   
- `1.4.0` Releases - [1.4.0](#140) | [1.4.1](#141) | [1.4.2](#142)   
- `1.3.0` Releases - [1.3.0](#130)   
- `1.2.0` Releases - [1.2.0](#120)   
- `1.1.0` Releases - [1.1.0](#110) | [1.1.1](#111) | [1.1.2](#112)  
- `1.0.0` Releases - [1.0.0](#100) | [1.0.1](#101) | [1.0.2](#102) | [1.0.3](#103)  
- `0.10.x` Releases - [0.10.0](#0100) | [0.10.1](#0101) | [0.10.2](#0102)  
- `0.9.x` Releases - [0.9](#09) | [0.9.1](#091) | [0.9.2](#092) | [0.9.3](#093) | [0.9.4](#094) | [0.9.5](#095) | [0.9.6](#096) | [0.9.7](#097) | [0.9.8](#098) | [0.9.9](#099) | [0.9.10](#0910) | [0.9.11](#0911) | [0.9.12](#0912) | [0.9.13](#0913) | [0.9.14](#0914)  
- `0.8.x` Releases - [0.8](#08)

---

## [3.5.2](https://github.com/andreamazz/AMPopTip/releases/tag/3.5.2)

- Removed previously added originating view and passing frame again. Added a boolean to consider the originating frame to trigger the tapHandler.

## [3.5.1](https://github.com/andreamazz/AMPopTip/releases/tag/3.5.1)

- Add a listener to the originating view.

## [3.5.0](https://github.com/andreamazz/AMPopTip/releases/tag/3.5.0)

- Swift 4.2 support

## [3.4.0](https://github.com/andreamazz/AMPopTip/releases/tag/3.4.0)

- Merged #166

## [3.3.0](https://github.com/andreamazz/AMPopTip/releases/tag/3.3.0)

- Merged #157
- Added support for shadows

## [3.2.1](https://github.com/andreamazz/AMPopTip/releases/tag/3.2.1)

Fixed the float animation  

## [3.2.0](https://github.com/andreamazz/AMPopTip/releases/tag/3.2.0)

- Merged #152

### Changed

The animation `float` now accepts two associated values: `float(offsetX: CGFloat?, offsetY: CGFloat?)`

## [3.1.4](https://github.com/andreamazz/AMPopTip/releases/tag/3.1.4)

- Merged #150

## [3.1.3](https://github.com/andreamazz/AMPopTip/releases/tag/3.1.3)

- Fixed #147

## [3.1.2](https://github.com/andreamazz/AMPopTip/releases/tag/3.1.2)

- Merged #146

## [3.1.1](https://github.com/andreamazz/AMPopTip/releases/tag/3.1.1)

- Merged #143

## [3.1.0](https://github.com/andreamazz/AMPopTip/releases/tag/3.1.0)

- Merged #142  

## [3.0.3](https://github.com/andreamazz/AMPopTip/releases/tag/3.0.3)

- Merged #135, #136, #137  

## [3.0.2](https://github.com/andreamazz/AMPopTip/releases/tag/3.0.2)

- Fixed animation issues

## [3.0.1](https://github.com/andreamazz/AMPopTip/releases/tag/3.0.1)

- Fixed #133  

## [3.0.0](https://github.com/andreamazz/AMPopTip/releases/tag/3.0.0)

- Swift 4 support  

## [2.1.7](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.7)

- Fixed #125

## [2.1.6](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.6)

- Merged #121

## [2.1.5](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.5)

- Fixed #114
- Fixed #119

## [2.1.4](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.4)

- Fixed issue with text rendering

## [2.1.3](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.3)

- Fixed an issue with the timer not being invalidated

## [2.1.2](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.2)

- Fixed regression with the iOS extensions

## [2.1.1](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.1)

- Fixed #117
- Fixed regression with the iOS extensions  

## [2.1.0](https://github.com/andreamazz/AMPopTip/releases/tag/2.1.0)

- Added ability to change the bubble contents    

## [2.0.1](https://github.com/andreamazz/AMPopTip/releases/tag/2.0.1)

- Fixed #116  

## [2.0.0](https://github.com/andreamazz/AMPopTip/releases/tag/2.0.0)

- AMPopTip was re-written in Swift. The API was also slightly changed. Make sure to check out the README.  

## [1.5.2](https://github.com/andreamazz/AMPopTip/releases/tag/1.5.2)

- Fixed #114  

## [1.5.1](https://github.com/andreamazz/AMPopTip/releases/tag/1.5.1)

- Merged #111  

## [1.5.0](https://github.com/andreamazz/AMPopTip/releases/tag/1.5.0)

- Added dimmable background (#107)  

## [1.4.2](https://github.com/andreamazz/AMPopTip/releases/tag/1.4.2)

- Fix #100

## [1.4.1](https://github.com/andreamazz/AMPopTip/releases/tag/1.4.1)

- Fix extensions handling with `NS_EXTENSION_UNAVAILABLE_IOS`

## [1.4.0](https://github.com/andreamazz/AMPopTip/releases/tag/1.4.0)

- Added `hideForced:`  

## [1.3.0](https://github.com/andreamazz/AMPopTip/releases/tag/1.3.0)

- Merged #87  

## [1.2.0](https://github.com/andreamazz/AMPopTip/releases/tag/1.2.0)

- Fix issue with UIFont set as appearance proxy.

## [1.1.2](https://github.com/andreamazz/AMPopTip/releases/tag/1.1.2)

- Merged #82

## [1.1.1](https://github.com/andreamazz/AMPopTip/releases/tag/1.1.1)

- Fixed #80

## [1.1.0](https://github.com/andreamazz/AMPopTip/releases/tag/1.1.0)

- Merged #75  

## [1.0.3](https://github.com/andreamazz/AMPopTip/releases/tag/1.0.3)

- Fixed #73

## [1.0.2](https://github.com/andreamazz/AMPopTip/releases/tag/1.0.2)

- Fixed #72

## [1.0.1](https://github.com/andreamazz/AMPopTip/releases/tag/1.0.1)

- Fixed #71 and #63

## [1.0.0](https://github.com/andreamazz/AMPopTip/releases/tag/1.0.0)

Stable release

## [0.10.2](https://github.com/andreamazz/AMPopTip/releases/tag/0.10.2)

#### Added
- Ability to define exit animations. Thanks to [@valeriomazzeo](https://github.com/valeriomazzeo)  


Source/PopTip+Transitions.swift
//
//  PopTip.swift
//  AMPopTip
//
//  Created by Andrea Mazzini on 01/05/2017.
//  Copyright © 2017 Andrea Mazzini. All rights reserved.
//

import UIKit

public extension PopTip {

  /// Triggers the chosen entrance animation
  ///
  /// - Parameter completion: the completion handler
  public func performEntranceAnimation(completion: @escaping () -> Void) {
    switch entranceAnimation {
    case .scale:
      entranceScale(completion: completion)
    case .transition:
      entranceTransition(completion: completion)
    case .fadeIn:
      entranceFadeIn(completion: completion)
    case .custom:
      if let backgroundMask = backgroundMask {
        containerView?.addSubview(backgroundMask)
      }
      containerView?.addSubview(self)
      entranceAnimationHandler?(completion)
    case .none:
      if let backgroundMask = backgroundMask {
        containerView?.addSubview(backgroundMask)
      }
      containerView?.addSubview(self)
      completion()
    }
  }

  /// Triggers the chosen exit animation
  ///
  /// - Parameter completion: the completion handler
  public func performExitAnimation(completion: @escaping () -> Void) {
    switch exitAnimation {
    case .scale:
      exitScale(completion: completion)
    case .fadeOut:
      exitFadeOut(completion: completion)
    case .custom:
      exitAnimationHandler?(completion)
    case .none:
      completion()
    }
  }

  private func entranceTransition(completion: @escaping () -> Void) {
    transform = CGAffineTransform(scaleX: 0.6, y: 0.6)
    switch direction {
    case .up:
      transform = transform.translatedBy(x: 0, y: -from.frame.origin.y)
    case .down, .none:
      transform = transform.translatedBy(x: 0, y: (containerView?.frame.height ?? 0) - from.frame.origin.y)
    case .left:
      transform = transform.translatedBy(x: from.frame.origin.x, y: 0)
    case .right:
      transform = transform.translatedBy(x: (containerView?.frame.width ?? 0) - from.frame.origin.x, y: 0)
    }
    if let backgroundMask = backgroundMask {
      containerView?.addSubview(backgroundMask)
    }
    containerView?.addSubview(self)

    UIView.animate(withDuration: animationIn, delay: delayIn, usingSpringWithDamping: 0.6, initialSpringVelocity: 1.5, options: [.curveEaseInOut, .beginFromCurrentState], animations: { 
      self.transform = .identity
      self.backgroundMask?.alpha = 1
    }) { (_) in
      completion()
    }
  }

  private func entranceScale(completion: @escaping () -> Void) {
    transform = CGAffineTransform(scaleX: 0, y: 0)
    if let backgroundMask = backgroundMask {
      containerView?.addSubview(backgroundMask)
    }
    containerView?.addSubview(self)

    UIView.animate(withDuration: animationIn, delay: delayIn, usingSpringWithDamping: 0.6, initialSpringVelocity: 1.5, options: [.curveEaseInOut, .beginFromCurrentState], animations: {
      self.transform = .identity
      self.backgroundMask?.alpha = 1
    }) { (_) in
      completion()
    }
  }

  private func entranceFadeIn(completion: @escaping () -> Void) {
    if let backgroundMask = backgroundMask {
      containerView?.addSubview(backgroundMask)
    }
    containerView?.addSubview(self)

    alpha = 0
    UIView.animate(withDuration: animationIn, delay: delayIn, options: [.curveEaseInOut, .beginFromCurrentState], animations: { 
      self.alpha = 1
      self.backgroundMask?.alpha = 1
    }) { (_) in
      completion()
    }
  }

  private func exitScale(completion: @escaping () -> Void) {
    transform = .identity

    UIView.animate(withDuration: animationOut, delay: delayOut, options: [.curveEaseInOut, .beginFromCurrentState], animations: { 
      self.transform = CGAffineTransform(scaleX: 0.0001, y: 0.0001)
      self.backgroundMask?.alpha = 0
    }) { (_) in
      completion()
    }
  }

  private func exitFadeOut(completion: @escaping () -> Void) {
    alpha = 1

    UIView.animate(withDuration: animationOut, delay: delayOut, options: [.curveEaseInOut, .beginFromCurrentState], animations: {
      self.alpha = 0
      self.backgroundMask?.alpha = 0
    }) { (_) in
      completion()
    }
  }
}


Demo/PopTip Demo/ViewController.swift
//
//  ViewController.swift
//  PopTip Demo
//
//  Created by Andrea Mazzini on 01/05/2017.
//  Copyright © 2017 Andrea Mazzini. All rights reserved.
//

import UIKit
import AMPopTip

class ViewController: UIViewController {

  enum ButtonType: Int {
    case center, topLeft, topRight, bottomLeft, bottomRight
  }

  let popTip = PopTip()
  var direction = PopTipDirection.up
  var topRightDirection = PopTipDirection.down
  var timer: Timer? = nil

  let /* Rival Sons's Tied Up */ lyrics = [
    "Go to the dark side full moon",
    "You shoot the apple off of my head",
    "'Cause your love, sweet love, is all that you put me through",
    "And honey without it you know I'd rather be dead",
    "I'm tied up",
    "I'm tangled up",
    "And I'm all wrapped up",
    "In you"
  ]

  override func viewDidLoad() {
    super.viewDidLoad()

    popTip.font = UIFont(name: "Avenir-Medium", size: 12)!
    popTip.shouldDismissOnTap = true
    popTip.shouldDismissOnTapOutside = true
    popTip.shouldDismissOnTapOriginatingView = true
    popTip.shouldDismissOnSwipeOutside = true
    popTip.edgeMargin = 5
    popTip.offset = 2
    popTip.bubbleOffset = 0
    popTip.edgeInsets = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
    
    /*
     Other customization:
    */
//    popTip.borderWidth = 2
//    popTip.borderColor = UIColor.blue
//    popTip.shadowOpacity = 0.4
//    popTip.shadowRadius = 3
//    popTip.shadowOffset = CGSize(width: 1, height: 1)
//    popTip.shadowColor = .black
    
    popTip.actionAnimation = .bounce(8)

    popTip.tapPopTipHandler = { _ in
      print("tap")
    }

    popTip.tapOutsideHandler = { _ in
      print("tap outside")
    }

    popTip.tapOriginatingViewHandler = { _ in
      print("tap on originating view")
    }

    popTip.swipeOutsideHandler = { _ in
      print("swipe outside")
    }

    popTip.dismissHandler = { _ in
      print("dismiss")
    }
  }

  @IBAction func action(sender: UIButton) {
    guard let button = ButtonType(rawValue: sender.tag) else { return }

    timer?.invalidate()
    
    popTip.arrowRadius = 0
    
    switch button {
    case .topLeft:
      let customView = UIView(frame: CGRect(x: 0, y: 0, width: 80, height: 120))
      let imageView = UIImageView(image: UIImage(named: "comment"))
      imageView.frame = CGRect(x: (80 - imageView.frame.width) / 2, y: 0, width: imageView.frame.width, height: imageView.frame.height)
      customView.addSubview(imageView)
      let label = UILabel(frame: CGRect(x: 0, y: imageView.frame.height, width: 80, height: 120 - imageView.frame.height))
      label.numberOfLines = 0
      label.text = "This is a custom view"
      label.textAlignment = .center
      label.textColor = .white
      label.font = UIFont.systemFont(ofSize: 12)
      customView.addSubview(label)
      popTip.bubbleColor = UIColor(red: 0.95, green: 0.65, blue: 0.21, alpha: 1)
      popTip.show(customView: customView, direction: .down, in: view, from: sender)

      popTip.entranceAnimationHandler = { [weak self] completion in
        guard let `self` = self else { return }
        self.popTip.transform = CGAffineTransform(rotationAngle: 0.3)
        UIView.animate(withDuration: 0.5, animations: { 
          self.popTip.transform = .identity
        }, completion: { (_) in
          completion()
        })
      }

    case .topRight:
      popTip.bubbleColor = UIColor(red: 0.97, green: 0.9, blue: 0.23, alpha: 1)
      if topRightDirection == .left {
        topRightDirection = .down
      } else {
        topRightDirection = .left
      }
      popTip.show(text: "I have a offset to move the bubble down or left side.", direction: topRightDirection, maxWidth: 150, in: view, from: sender)
    case .bottomLeft:
      popTip.bubbleColor = UIColor(red: 0.73, green: 0.91, blue: 0.55, alpha: 1)
      let attributes: [NSAttributedString.Key: Any] = [NSAttributedString.Key.font: UIFont.systemFont(ofSize: 12), NSAttributedString.Key.foregroundColor: UIColor.white]
      let underline: [NSAttributedString.Key: Any] = [NSAttributedString.Key.font: UIFont.systemFont(ofSize: 13), NSAttributedString.Key.foregroundColor: UIColor.white, NSAttributedString.Key.underlineStyle: NSUnderlineStyle.single.rawValue]
      let attributedText = NSMutableAttributedString(string: "I'm presenting a string ", attributes: attributes)
      attributedText.append(NSAttributedString(string: "with attributes!", attributes: underline))
      popTip.show(attributedText: attributedText, direction: .up, maxWidth: 200, in: view, from: sender)
    case .bottomRight:
      popTip.bubbleColor = UIColor(red: 0.81, green: 0.04, blue: 0.14, alpha: 1)
      popTip.show(text: "Animated popover, great for subtle UI tips and onboarding", direction: .left, maxWidth: 200, in: view, from: sender)
    case .center:
      popTip.arrowRadius = 2
      popTip.bubbleColor = UIColor(red: 0.31, green: 0.57, blue: 0.87, alpha: 1)
      popTip.show(text: "Animated popover, great for subtle UI tips and onboarding", direction: direction, maxWidth: 200, in: view, from: sender)
      direction = direction.cycleDirection()
      timer = Timer.scheduledTimer(withTimeInterval: 2, repeats: true) { (_) in
        self.popTip.update(text: self.lyrics.sample())
      }
    }
  }

}

extension PopTipDirection {
  func cycleDirection() -> PopTipDirection {
    switch self {
    case .up:
      return .right
    case .right:
      return .down
    case .down:
      return .left
    case .left:
      return .up
    case .none:
      return .none
    }
  }
}

extension Array {
  func sample() -> Element {
    let index = Int(arc4random_uniform(UInt32(count)))
    return self[index]
  }
}
